### ISSUE TITLE
? operator gives wrong results

### ISSUE THREAD

**Issue by ericsagnes** [NixOS/nix#1059]
The `?` operator to check if a set has a key gives a wrong result when the key name is a passed via a function parameter:

```
nix-repl> let s = { a = 1; }; in [ (s ? "a") ((x: s ? x) "a") ]
[ true false ]
```

**Comment by ericsagnes** [comment]
Found a workaround, when passed via a function parameter, the key name must be escaped with `${}`.

```
nix-repl> let s = { a = 1; }; in [ (s ? "a") ((x: s ? ${x}) "a") ]
[ true true ]
```

**Comment by domenkozar** [comment]
We should document this better in the reference.

**Comment by edolstra** [comment]
Closing as not a bug. `s ? x` means testing for the attribute literally named `x`. This is similar to how `s . x` selects the attribute named `x`.

**Comment by domenkozar** [comment]
I'd like to keep this open until we're sure documentation states it clearly enough.

**Comment by ericsagnes** [comment]
I know that Nix expression language is not lambda calculus, but assuming that η-conversion holds in a functional language seems pretty legit.

`s ? "a"` is expected to be equivalent to `(x: s ? x) "a"` in the same way that `(x: 1 + x) 1` is expected to be equivalent to `1 + 1`.

It took me quite some debug time to find out why a simple check function was returning wrong results.

As far as understand it, η-conversion doesn't hold because the `?` operator has an "effect" on its righthand value.
Nix expects the righthand to be an `attrpath`, so if it is a string it will automatically converts it to an `attrpath`.
If the righthand is a reference to a string, Nix will not try to transform it into an `attrpath`, and the check will silently fails.

Examples:

- normal use case, works as expected:

```
> let s = { a = 1; }; in  s ? a
true
```

- normal use case with a string, works as expected:

```
> let s = { a = 1; }; in  s ? "a"
true
```

- using a function parameter reference with an attrpath, fails with error

```
> let s = { a = 1; }; in  (x: s ? x) a
undefined variable ‘a’ at (string):1:36
```

- using a function parameter reference with a string, silently fails

```
let s = { a = 1; }; in  (x: s ? x) "a"
false
```

- using a let reference as an attrpath, fails with error:

```
> let x = a; s = { a = 1; }; in s ? x
error: undefined variable ‘a’ at (string):1:9
```

- using a let reference as a string, silently fails:

```
> let x = "a"; s = { a = 1; }; in s ? x
false
```

IMHO, the cases where `?` silently fails are bad, and I would prefer:
- a failure with error
- or correct behavior with string references

TL;DR Having the `?` operator giving wrong results silently depending on its usage is just calling for trouble.

**Comment by ericsagnes** [comment]
Just a thought, but introducing a dedicated notation for `attrpaths` could be an interesting way to solve this issue. 
And that could be useful for other purposes as attribute sets are pretty central in the Nix expression language.

**Comment by edolstra** [comment]
η-conversion doesn't really have anything to do with, because the ?-operator is a special syntactical construct (just like the dot operator) - it's not a infix binary function.

BTW, this operator has had this meaning for the last 12 years so it's a bit too late to change it :-)

**Comment by domenkozar** [comment]
Yeah:

```
$ cd ~/dev/nixpkgs
$ git grep "? [a-z]" | wc -l
3860
```

**Comment by ericsagnes** [comment]
OK, I am convinced :)

I opened a PR on the documentation that adds references to alternative functions that fixes the issue.

**Comment by roberth** [comment]
I think we should promote `x?a` as the preferred formatting for `?`, as it is more similar to the formatting everyone uses with `.`.

It's a simple thing we could do to nudge the balance just a bit towards the correct interpretation of this operator.

To formalize the formatting rule:
 - Use a space on both sides when the operator is algebraic within the set of values: takes two values to return a value.
 - Use no whitespace when the operator is not algebraic. `.` and `?` both take an identifier as the second value, which represents a name, not a value.

This is arguably a semantically more relevant distinction than "use `.` like you do in other languages, and all the other operators use spaces" which seems to have been the convention, at least in Nixpkgs.

```nix
# attribute operators
a.attr
a?attr

# algebraic operators within value
a + b
a ++ b
a < b
# ...
```

**Comment by roberth** [comment]
Formatting "hasAttr" as `x?a` without whitespace has another advantage, which is to visually differentiate it from the other question mark used in parameter defaults, `{ x ? a }: x`.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
